packages/sitecore-jss-react/src/components/BYOCRenderer.tsx

```diff
@@ -0,0 +1,112 @@
+import React from 'react';
+import { ComponentFields } from '@sitecore-jss/sitecore-jss/layout';
+import { getDataFromFields } from '../utils';
+import { MissingComponent } from './MissingComponent';
+import { RegisteredComponents } from '@sitecore-feaas/clientside/types/ui/FEAASExternal';
+
+/**
+ * Data from rendering params on Sitecore's BYOC rendering
+ */
+type BYOCRenderingParams = {
+  /**
+   * Name of the component to render
+   */
+  ComponentName: string;
+  /**
+   * JSON props to pass into rendered component
+   */
+  ComponentProps?: string;
+  styles?: string;
+  RenderingIdentifier?: string;
+};
+
+/**
+ * Props for BYOC wrapper component
+ */
+export type BYOCProps = {
+  /**
+   * rendering params
+   */
+  params?: BYOCRenderingParams;
+  /**
+   * fields from datasource items to be passed as rendered child component props
+   */
+  fields?: ComponentFields;
+};
+
+/**
+ * Props for BYOCRenderer component. Includes components list to load external components from.
+ */
+type BYOCRendererProps = BYOCProps & {
+  components: RegisteredComponents;
+};
+
+/**
+ * BYOCRenderer helps rendering BYOC components - that can be taken from anywhere
+ * and registered without being deployed as Sitecore renderings
+ * @param {ByocRendererProps} props component props
+ * @returns dynamicly rendered component or Missing Component error frame
+ */
+export class BYOCRenderer extends React.Component<BYOCRendererProps> {
+  state: Readonly<{ error?: Error }>;
+
+  constructor(props: BYOCRendererProps) {
+    super(props);
+    this.state = {};
+  }
+
+  static getDerivedStateFromError(error: Error) {
+    // Update state so the next render will show the fallback UI.
+    return { error: error };
+  }
+
+  componentDidCatch(error: Error) {
+    this.setState({ error });
+  }
+
+  render() {
+    const props: BYOCRendererProps = this.props;
+    if (this.state.error) {
+      return <div>A rendering error occurred: {this.state.error.message}.</div>;
```

Comment:
What do you think about adding an optional `errorComponent` to the `BYOCRendererProps` and using this for any errors if provided? We might also consider adding a `missingComponentComponent` override as well. This would provide flexibility for customers to be able to override this markup/message. Basically, following the pattern of what we provide on the [Placeholder component](https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx#L67-L82).